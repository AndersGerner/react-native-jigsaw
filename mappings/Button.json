[{"name":"Button Outline","tag":"Button","category":"button","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Button_Outline.png","props":{"icon":{"label":"Icon Name","description":"Name of the icon","editable":true,"required":true,"type":"icon","value":null},"children":{"label":"Label","description":"Button label","required":true,"editable":true,"type":"string","value":"Get Started"},"color":{"label":"Color Override","description":"Override the background color of the button","editable":true,"required":false,"type":"color","value":null},"labelColor":{"label":"Label Color Override","description":"Override the label color of the button","editable":true,"required":false,"type":"color","value":null},"disabled":{"label":"Disabled","description":"Whether the button should be disabled","editable":true,"required":false,"type":"boolean","value":null},"loading":{"label":"Loading","description":"Whether to show a loading indicator","editable":true,"required":false,"type":"boolean","value":null},"onPress":{"label":"Action","description":"Action to execute when button pressed","editable":true,"type":"action","value":null},"type":{"label":"Type","description":"Button type","editable":false,"required":true,"type":"string","value":"outline"}},"layout":{}},{"name":"Button Solid","tag":"Button","category":"button","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Button_Solid.png","props":{"icon":{"label":"Icon Name","description":"Name of the icon","editable":true,"required":true,"type":"icon","value":null},"children":{"label":"Label","description":"Button label","required":true,"editable":true,"type":"string","value":"Get Started"},"color":{"label":"Color Override","description":"Override the background color of the button","editable":true,"required":false,"type":"color","value":null},"labelColor":{"label":"Label Color Override","description":"Override the label color of the button","editable":true,"required":false,"type":"color","value":null},"disabled":{"label":"Disabled","description":"Whether the button should be disabled","editable":true,"required":false,"type":"boolean","value":null},"loading":{"label":"Loading","description":"Whether to show a loading indicator","editable":true,"required":false,"type":"boolean","value":null},"onPress":{"label":"Action","description":"Action to execute when button pressed","editable":true,"type":"action","value":null},"type":{"label":"Type","description":"Button type","editable":false,"required":true,"type":"string","value":"solid"}},"layout":{}}]