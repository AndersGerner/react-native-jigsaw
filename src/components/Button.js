/* @flow */

import * as React from "react";
import {
  ActivityIndicator,
  Animated,
  View,
  Text,
  StyleSheet
} from "react-native";
import color from "color";
import Config from "./Config";
import Icon from "./Icon";
import Touchable from "./Touchable";
import Elevation from "./Elevation";
import { withTheme } from "../core/theming";
import type { Theme } from "../types";
import type { IconSource } from "./Icon";
import { FORM_TYPES, COMPONENT_TYPES } from "../core/component-types";

type Props = {
  /**
   * Type of the button. You can change the type to adjust the styling to give it desired emphasis.
   * - `text` - flat button without background or outline (low emphasis)
   * - `outline` - button with an outline (medium emphasis)
   * - `solid` - button with a background color (high emphasis)
   */
  type?: "text" | "outline" | "solid",
  /**
   * Custom text color for flat button, or background color for solid button.
   */
  color?: string,
  /**
   * Whether to show a loading indicator.
   */
  loading?: boolean,
  /**
   * Icon to display for the `Button`.
   */
  icon?: IconSource,
  /**
   * Whether the button is disabled. A disabled button is greyed out and `onPress` is not called on touch.
   */
  disabled?: boolean,
  /**
   * Label text of the button.
   */
  children: React.Node,
  /**
   * Function to execute on press.
   */
  onPress?: () => mixed,
  style?: any,
  /**
   * @optional
   */
  theme: Theme
};

/**
 * A button is component that the user can press to trigger an action.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img src="screenshots/button-1.png" />
 *     <figcaption>Text button</figcaption>
 *   </figure>
 *   <figure>
 *     <img src="screenshots/button-2.png" />
 *     <figcaption>Outlined button</figcaption>
 *   </figure>
 *   <figure>
 *     <img src="screenshots/button-3.png" />
 *     <figcaption>Contained button</figcaption>
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Button } from '@draftbit/ui';
 *
 * const MyComponent = () => (
 *   <Button icon="add-a-photo" type="solid" onPress={() => console.log('Pressed')}>
 *     Press me
 *   </Button>
 * );
 *
 * export default MyComponent;
 * ```
 */

class Button extends React.Component<Props> {
  static defaultProps = {
    elevation: 0,
    type: "solid"
  };

  render() {
    const {
      disabled,
      type,
      loading,
      icon,
      color: colorOverride,
      children,
      onPress,
      elevation,
      style,
      theme,
      ...rest
    } = this.props;

    const {
      colors,
      disabledOpacity,
      borderRadius,
      spacing,
      typography
    } = theme;

    let backgroundColor, borderColor, textColor, borderWidth;
    const buttonColor = colorOverride || colors.primary;

    if (type === "solid") {
      backgroundColor = buttonColor;

      if (disabled) {
        textColor = color(colors.surface)
          .alpha(disabledOpacity)
          .rgb()
          .string();
      } else {
        textColor = colors.surface;
      }
    } else {
      backgroundColor = "transparent";

      if (disabled) {
        textColor = color(buttonColor)
          .alpha(disabledOpacity)
          .rgb()
          .string();
      } else {
        textColor = buttonColor;
      }
    }

    if (type === "outline") {
      if (disabled) {
        borderColor = color(buttonColor)
          .alpha(disabledOpacity)
          .rgb()
          .string();
      } else {
        borderColor = buttonColor;
      }
      borderWidth = StyleSheet.hairlineWidth;
    } else {
      borderColor = "transparent";
      borderWidth = 0;
    }

    const buttonStyle = {
      backgroundColor,
      borderColor,
      borderWidth,
      borderRadius: borderRadius.button
    };

    const textStyle = {
      textAlign: "center",
      color: textColor,
      marginVertical: spacing.large,
      marginHorizontal: spacing.large
    };

    const iconStyle = [
      styles.icon,
      {
        marginLeft: spacing.large,
        marginRight: -8,
        width: Config.buttonIconSize
      }
    ];

    return (
      <Elevation style={{ elevation, alignSelf: "stretch" }}>
        <Touchable
          {...rest}
          onPress={onPress}
          accessibilityTraits={disabled ? ["button", "disabled"] : "button"}
          accessibilityComponentType="button"
          disabled={disabled || loading}
          style={[styles.button, buttonStyle, style]}
        >
          <View style={styles.content}>
            {icon && loading !== true ? (
              <View style={iconStyle}>
                <Icon
                  name={icon}
                  size={Config.buttonIconSize}
                  color={textColor}
                />
              </View>
            ) : null}
            {loading ? (
              <ActivityIndicator
                size="small"
                color={textColor}
                style={iconStyle}
              />
            ) : null}
            <Text
              numberOfLines={1}
              style={[styles.label, textStyle, typography.button]}
            >
              {children}
            </Text>
          </View>
        </Touchable>
      </Elevation>
    );
  }
}

const styles = StyleSheet.create({
  button: {
    minWidth: 64,
    borderStyle: "solid"
  },
  content: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center"
  },
  icon: {
    width: Config.buttonIconSize
  }
});

export default withTheme(Button);

const SEED_DATA_PROPS = {
  icon: {
    label: "Icon Name",
    description: "Name of the icon",
    editable: true,
    required: true,
    type: FORM_TYPES.icon,
    value: "MaterialIcons/add"
  },
  children: {
    label: "Label",
    description: "Button label",
    required: true,
    editable: true,
    type: FORM_TYPES.string,
    value: "Get Started"
  },
  color: {
    label: "Color Override",
    description: "Override the background color of the button",
    editable: true,
    required: false,
    type: FORM_TYPES.color,
    value: null
  },
  disabled: {
    label: "Disabled",
    description: "Whether the button should be disabled",
    editable: true,
    required: false,
    type: FORM_TYPES.boolean,
    value: null
  },
  loading: {
    label: "Loading",
    description: "Whether to show a loading indicator",
    editable: true,
    required: false,
    type: FORM_TYPES.boolean,
    value: null
  }
};

export const SEED_DATA = [
  {
    name: "Button Outline",
    tag: "Button",
    category: COMPONENT_TYPES.button,
    preview_image_url:
      "{CLOUDINARY_URL}/Button_Outline.png",
    props: {
      ...SEED_DATA_PROPS,
      type: {
        label: "Type",
        description: "Button type",
        editable: false,
        required: true,
        type: FORM_TYPES.string,
        value: "outline"
      }
    },
    layout: {
      width: 343,
      height: 48
    }
  },
  {
    name: "Button Solid",
    tag: "Button",
    category: COMPONENT_TYPES.button,
    preview_image_url:
      "{CLOUDINARY_URL}/Button_Solid.png",
    props: {
      ...SEED_DATA_PROPS,
      type: {
        label: "Type",
        description: "Button type",
        editable: false,
        required: true,
        type: FORM_TYPES.string,
        value: "solid"
      }
    },
    layout: {
      width: 343,
      height: 48
    }
  },
  {
    name: "Button Text",
    tag: "Button",
    category: COMPONENT_TYPES.button,
    preview_image_url:
      "{CLOUDINARY_URL}/Button_Text.png",
    props: {
      ...SEED_DATA_PROPS,
      type: {
        label: "Type",
        description: "Button type",
        editable: false,
        required: true,
        type: FORM_TYPES.string,
        value: "text"
      }
    },
    layout: {
      width: 343,
      height: 48
    }
  }
];
