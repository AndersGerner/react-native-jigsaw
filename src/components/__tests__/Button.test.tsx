import * as React from "react";
import renderer from "react-test-renderer";
import Button from "../Button";

it("renders button", () => {
  const tree = renderer.create(<Button onPress={() => {}} />).toJSON();

  expect(tree).toMatchSnapshot();
});
