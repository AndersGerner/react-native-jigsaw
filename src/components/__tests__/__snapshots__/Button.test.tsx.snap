// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "elevation": 0,
      "zIndex": 100,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(90, 69, 255, 1)",
        "borderColor": "transparent",
        "borderRadius": 24,
        "borderStyle": "solid",
        "borderWidth": 0,
        "minWidth": 64,
        "opacity": 1,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "rgba(255, 255, 255, 1)",
                "marginHorizontal": 16,
                "marginVertical": 16,
                "textAlign": "center",
              },
              Object {
                "backgroundColor": "transparent",
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 16,
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
